### PR TITLE
Microsoft teams udf

### PR DESCRIPTION
**Pull request by mahara13**

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -1,4 +1,4 @@
-rootProject.name = ''
+rootProject.name = 'overops-functions'
```
Comment (dor-levi) [inline settings.gradle]: Why is this needed?

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextSection.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.takipi.udf.microsoftteams.card;
+
+public class MicrosoftTextSection implements MicrosoftSection {
+    private String text = "";
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextSection.java]: This is only used when serializing as json, so IDEs mark it as unused with a warning.
Solution, make it `public final`
As a `String` object is immutable, there's no risk of it being internally modified even if final.

Also, as a rule of thumb, all members that _can_ be `final`, _should_ be `final`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextBuilder.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        return this;
+    }
+
+    public MicrosoftTextBuilder addEnter() {
+        stringBuilder.append("\n\n");
+
+        return this;
+    }
+
+    public String build() {
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextBuilder.java]: It's easier to look at a builder object when the `build` method is the last (non-static) method in the class.

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextBuilder.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        return stringBuilder.toString();
+    }
+
+
+    public MicrosoftTextBuilder addArray(List<Location> stack_frames, String prefix) {
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextBuilder.java]: While the JSON is being serialized with underscores, method parameters should be in camel case (i.e `stackFrames` in this case)

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTarget.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.takipi.udf.microsoftteams.card;
+
+public class MicrosoftTarget {
+    public String os = "default";
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTarget.java]: Both members can probably be `final`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftPotentialAction.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            return this;
+        }
+
+        public Builder setTargets(List<MicrosoftTarget> targets) {
+            this.targets = targets;
+            return this;
+        }
+
+        public Builder addTargets(MicrosoftTarget... targets) {
+            if (this.targets == null) this.targets = new ArrayList<>();
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftPotentialAction.java]: Please have `{}` surrounding `if` code blocks, even if it's just a single line.
In this case:
```
if (this.targets == null) {
    this.targets = new ArrayList<>();
}
```

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftPotentialAction.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.google.gson.annotations.SerializedName;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class MicrosoftPotentialAction {
+    @SerializedName("@type")
+    public String type;
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftPotentialAction.java]: Members of this class can probably be `final`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftCard.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            return this;
+        }
+
+        public Builder setSections(List<MicrosoftSection> sections) {
+            this.sections = sections;
+            return this;
+        }
+
+        public Builder addSections(MicrosoftSection... sections) {
+            if (this.sections == null) this.sections = new ArrayList<>();
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftCard.java]: Please have {} surrounding if code blocks, even if it's just a single line.
In this case:

```
if (this.sections == null) {
    this.sections = new ArrayList<>();
}
```

Same for other methods in this class

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftActivitySection.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            return this;
+        }
+
+        public Builder setFacts(List<MicrosoftFact> facts) {
+            this.facts = facts;
+            return this;
+        }
+
+        public Builder addFacts(MicrosoftFact... facts) {
+            if (this.facts == null) this.facts = new ArrayList<>();
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftActivitySection.java]: Please have {} surrounding if code blocks, even if it's just a single line.
In this case:

```
if (this.facts == null) {
    this.facts = new ArrayList<>();
}
```

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftCard.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
+package com.takipi.udf.microsoftteams.card;
+
+import com.google.gson.annotations.SerializedName;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class MicrosoftCard {
+    @SerializedName("@type")
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftCard.java]: Members of this class can probably be `final`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftActivitySection.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
+package com.takipi.udf.microsoftteams.card;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class MicrosoftActivitySection implements MicrosoftSection {
+    public String activityTitle = "";
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftActivitySection.java]: Members of this class can probably be `final`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsUtil.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+    public static void logUDFInput(String rawContextArgs, String rawInput) {
+        System.out.println("rawContextArgs");
+        System.out.println(rawContextArgs);
+        System.out.println("rawInput");
+        System.out.println(rawInput);
+    }
+
+    public static class TimeSlot {
+        public DateTime to;
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsUtil.java]: Members of this class can probably be `final`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsInput.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    public String toString() {
+        StringBuilder builder = new StringBuilder();
+
+        builder.append("Microsoft Teams Alert ( url = ").append(url).append(" )");
+
+        return builder.toString();
+    }
+
+    public void checkUrl() {
+        if (url == null || !url.matches("^https.*outlook.*webhook.*")) {
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsInput.java]: Please have each if condition in it's own parentheses.
i.e. in this case:
```
if ((url == null) || (!url.matches("^https.*outlook.*webhook.*"))) {
```

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelRequest.java [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class MicrosoftTeamsChannelRequest implements ApiPostRequest<EmptyResult> {
+
+    public static final String THEME_COLOR = "ffc000";
+    private String url;
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelRequest.java]: Members of this class can probably be `final`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+        // set url similar to "url=https://outlook.office.com/webhook/..."
+        String rawInput = "url=https://outlook.office.com/webhook/...";
+        // convert context args to a JSON string
+        String rawContextArgs = new Gson().toJson(contextArgs);
+
+        // execute the UDF
+        MicrosoftTeamsChannelFunction.execute(rawContextArgs, rawInput);
+    }
+
+    public static ContextArgs getContextArgs(String[] args) {
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: This method should be in `TestUtil`, or utilize on of the existing methods in there

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+    private static UrlClient.Response<EventResult> getEventResultResponse(ContextArgs args, ApiClient apiClient) {
+        EventRequest eventRequest = EventRequest.newBuilder()
+                .setEventId(args.eventId)
+                .setServiceId(args.serviceId)
+                .setIncludeStacktrace(true)
+                .build();
+
+        UrlClient.Response<EventResult> eventResultResponse = apiClient.get(eventRequest);
+
+        if (!eventResultResponse.isOK() || eventResultResponse.data == null)
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: Please have `{}` surrounding if code blocks, even if it's just a single line, as well as surround each `if` case in it's own parentheses
In this case:

```
if ((!eventResultResponse.isOK()) || (eventResultResponse.data == null)) {
    throw new IllegalStateException("Event is not present or api issue. eventId = " + args.eventId
                    + " , serviceId = " + args.serviceId);
}
```

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+                .setServer(eventsGraph.machine_name)
+                .setApplication(eventsGraph.application_name)
+                .setEnvironmentName(getEnvironmentName(eventData.environments, args.serviceId))
+                .setDeployment(eventsGraph.deployment_name)
+                .setDoNotAlertLink(getDoNotAlertLink(args, eventResultResponse.data.type))
+                .build();
+        UrlClient.Response<String> post = SimpleUrlClient.newBuilder().build().post(microsoftTeamsChannelRequest);
+        System.out.println("Post Microsoft Teams Webhook Channel request");
+
+        if (post.isBadResponse())
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: Please have `{}` surrounding if code blocks, even if it's just a single line,.
In this case:

```
if (post.isBadResponse()) {
    throw new IllegalStateException("Can't send card to " + input.url);
}
```

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+            e.printStackTrace();
+        }
+    }
+
+    private static void executeImplementation(String rawContextArgs, String rawInput) {
+        logUDFInput(rawContextArgs, rawInput);
+        MicrosoftTeamsInput input = getInput(rawInput);
+        ContextArgs args = (new Gson()).fromJson(rawContextArgs, ContextArgs.class);
+
+        if (!args.validate())
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: Please have `{}` surrounding if code blocks, even if it's just a single line,.
In this case:

```
if (!args.validate()) {
    throw new IllegalArgumentException("Bad context args: " + rawContextArgs);
}
if (!args.eventValidate()) {
    return;
}
```

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsAnomalyRequest.java [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class MicrosoftTeamsAnomalyRequest implements ApiPostRequest<EmptyResult> {
+
+    public static final String SUMMARY_IMAGE = "https://s3.amazonaws.com/www.takipi.com/email/v5/threshold-thumb.png";
+
+    private String url;
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsAnomalyRequest.java]: All members of this class can probably be `final`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsAnomalyFunction.java [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+        long hitSum = 0;
+
+        if (volumeResponse.data != null && volumeResponse.data.events != null) {
+            hitSum = volumeResponse.data.events.stream().map(v -> v.stats.hits).reduce(0L, (sum, c) -> sum + c);
+        }
+
+        return hitSum;
+    }
+
+    private static String getViewName(ContextArgs args, ApiClient apiClient, String viewId) {
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsAnomalyFunction.java]: Check `ViewUtil.getServiceView`

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsUtil.java [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import java.util.List;
+
+public class MicrosoftTeamsUtil {
+
+    public static EventData getEventData(ApiClient apiClient, ContextArgs args) {
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsUtil.java]: I don't understand the purpose of this class.
It seems like `servers`/`applications`/`deployments` are only used in the context of `EventUtil.getEventRecentLink`

However, giving that method all of them is the equivalent of giving that method none of them.
If none are passed, the api server won't filter the snapshot, which is the same as asking it to filter based on all of them.

Also, the process of calling `ClientUtil.getServers`, `ClientUtil.getApplications` and `ClientUtil.getDeployments` is time consuming for the UDF, and resource consuming for the API server.

Regarding the environments, there's now `ClientUtil.getEnvironment(apiClient, serviceId)` which can be used.

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+        UrlClient.Response<EventResult> eventResultResponse = getEventResultResponse(args, apiClient);
+        System.out.println("Got exceptionClassName, exception type, exceptionLocationPath, stack_frames, deployment");
+        MicrosoftTeamsUtil.EventData eventData = getEventData(apiClient, args);
+        System.out.println("Got applications, environments, deployments and servers");
+        MicrosoftTeamsUtil.TimeSlot timeSlot = getTimeSlot(MINUTES_TIME_SPAN);
+
+        SummarizedView view = getSummarizedView(rawContextArgs, args, apiClient);
+        System.out.println("Got viewId");
+
+        Graph eventsGraph = ViewUtil.getEventsGraph(apiClient, args.serviceId, view.id,
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: What is this graph used for?
Getting a graph is one of the most expensive operations possible, and it seems this is only used to get the server/application/deployment?
For what purpose?

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsAnomalyFunction.java [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static void executeImplementation(String rawContextArgs, String rawInput) {
+        logUDFInput(rawContextArgs, rawInput);
+        MicrosoftTeamsAnomalyInput input = getInput(rawInput);
+        ContextArgs args = (new Gson()).fromJson(rawContextArgs, ContextArgs.class);
+        if (!args.validate())
```
Comment (dor-levi) [inline overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsAnomalyFunction.java]: Please have `{}` surrounding if code blocks, even if it's just a single line. Also, this should probably check `viewValidate`
In this case:

```
if (!args.validate()) {
    throw new IllegalArgumentException("Bad context args: " + rawContextArgs);
}
if (!args.viewValidate()) {
    return;
}
```

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsAnomalyFunction.java (hunk 20 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsAnomalyFunction.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java (hunk 16 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java (hunk 15 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: Add builder for more flexibility and util method to create a default configuration

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelRequest.java (hunk 14 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelRequest.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsInput.java (hunk 13 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsInput.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsUtil.java (hunk 21 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsUtil.java]: Done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsUtil.java (hunk 12 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsUtil.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftActivitySection.java (hunk 11 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftActivitySection.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftCard.java (hunk 10 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftCard.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftCard.java (hunk 8 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftCard.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftPotentialAction.java (hunk 7 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftPotentialAction.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftPotentialAction.java (hunk 6 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftPotentialAction.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTarget.java (hunk 5 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTarget.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextBuilder.java (hunk 3 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextBuilder.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextBuilder.java (hunk 4 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextBuilder.java]: done, this class will not be serialized it just until class to create MicrosoftTeams Card text pseudo Html markup

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextSection.java (hunk 2 above)
Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/card/MicrosoftTextSection.java]: done

File: overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java (hunk 22 above)
Comment (dor-levi) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: I'll refine my Q -
What are the machine/application/deployment names *should* be?
The ones relating to the first snapshot ever for this event?
The ones relating to all the places this event happened so far?

Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: Here screenshots where related data I could find in the backend
image
image
image


In Microsoft Teams they will be reflected
image

Comment (dor-levi) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: OK
Starting with api-client 2.29.0, you'll have the following fields on `EventResult` -
`introduced_by_server` - the server in which this event first happened.
`introduced_by_application` - the application in which this event first happened.

The matching field for deployments is called `introduced_by`, and it's been there from earlier versions.

Will this solve your issue?

NOTE - while the fields themselves exist, the backend will only start populating them in a few days.
I just wanted to make sure that you can easily use a version that will allow you to work :)

Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: @dor-levi thanks, I will apply when 2.29.0 version will be available on prod

Comment (mahara13) [reply overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java]: @dor-levi  done
